satellite/satellitedb/gracefulexit.go

```diff
@@ -0,0 +1,232 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"context"
+
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite/gracefulexit"
+	dbx "storj.io/storj/satellite/satellitedb/dbx"
+)
+
+type gracefulexitDB struct {
+	db *dbx.DB
+}
+
+// CreateProgress creates a graceful exit progress entry in the database.
+func (db *gracefulexitDB) CreateProgress(ctx context.Context, nodeID storj.NodeID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	return db.db.CreateNoReturn_GracefulExitProgress(
+		ctx,
+		dbx.GracefulExitProgress_NodeId(nodeID.Bytes()),
+		dbx.GracefulExitProgress_BytesTransferred(0),
+	)
+}
+
+// UpdateProgress updates a graceful exit progress entry in the database.
+func (db *gracefulexitDB) UpdateProgress(ctx context.Context, nodeID storj.NodeID, bytesTransferred int64) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	return db.db.UpdateNoReturn_GracefulExitProgress_By_NodeId(
+		ctx,
+		dbx.GracefulExitProgress_NodeId(nodeID.Bytes()),
+		dbx.GracefulExitProgress_Update_Fields{
+			BytesTransferred: dbx.GracefulExitProgress_BytesTransferred(bytesTransferred),
+		},
+	)
+}
+
+// DeleteProgress deletes a graceful exit progress entry in the database.
+func (db *gracefulexitDB) DeleteProgress(ctx context.Context, nodeID storj.NodeID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	_, err = db.db.Delete_GracefulExitProgress_By_NodeId(ctx, dbx.GracefulExitProgress_NodeId(nodeID.Bytes()))
+	return err
+}
+
+// IncrementProgressBytesTransferred increments bytes transferred value.
+func (db *gracefulexitDB) IncrementProgressBytesTransferred(ctx context.Context, nodeID storj.NodeID, bytesTransferred int64) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	statement := db.db.Rebind(
+		`UPDATE graceful_exit_progress SET bytes_transferred = bytes_transferred + ? WHERE node_id = ?`,
+	)
+	_, err = db.db.ExecContext(ctx, statement, uint64(bytesTransferred), nodeID)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetProgress gets a graceful exit progress entry in the database.
+func (db *gracefulexitDB) GetProgress(ctx context.Context, nodeID storj.NodeID) (_ *gracefulexit.Progress, err error) {
+	defer mon.Task()(&ctx)(&err)
+	dbxProgress, err := db.db.Get_GracefulExitProgress_By_NodeId(ctx, dbx.GracefulExitProgress_NodeId(nodeID.Bytes()))
+	if err != nil {
+		return nil, err
+	}
+	nID, err := storj.NodeIDFromBytes(dbxProgress.NodeId)
+	if err != nil {
+		return nil, err
+	}
+
+	progress := &gracefulexit.Progress{
+		NodeID:           nID,
+		BytesTransferred: dbxProgress.BytesTransferred,
+		UpdatedAt:        dbxProgress.UpdatedAt,
+	}
+
+	return progress, err
+}
+
+// GetAllProgress gets all graceful exit progress entries in the database.
+func (db *gracefulexitDB) GetAllProgress(ctx context.Context) (_ []*gracefulexit.Progress, err error) {
```

Comment:
in what scenarios will we be using GetAllProgress? It seems like this could be an expensive operation.